easybuild/framework/easyblock.py

```diff
@@ -663,6 +667,52 @@ def obtain_file(self, filename, extension=False, urls=None, download_filename=No
                 return foundfile
+            elif git_config:
+                # if a non-empty dictionary was provided, download from git and archive
+                if not isinstance(git_config, dict):
+                    raise EasyBuildError("Found a non null git_config in 'sources', but value is not a dictionary")
+                else:
+                    git_config = git_config.copy()
+                    tag = git_config.pop('tag', None)
+                    url = git_config.pop('url', None)
+                    repo_name = git_config.pop('repo_name', None)
+                    commit = git_config.pop('commit', None)
+                    recursive = git_config.pop('recursive', False)
+                    if git_config:
+                        raise EasyBuildError("Found one or more unexpected keys in 'git_config' specification: %s",
+                                             git_config)
+                    if not repo_name:
+                        raise EasyBuildError("repo_name not specified in git_config parameter")
+                    if not tag and not commit:
+                        raise EasyBuildError("Neither tag nor commit found in git_config parameter")
+                    if tag and commit:
+                        raise EasyBuildError("Tag and commit are mutually exclusive in git_config parameter")
+                    if not url:
+                        raise EasyBuildError("url not specified in git_config parameter")
+                    if '.tar.gz' not in filename:
+                        raise EasyBuildError("git_config only supports filename ending in .tar.gz")
+
+                    targetdir = os.path.join(srcpaths[0], self.name.lower()[0], self.name)
+                    mkdir(targetdir, parents=True)
+                    targetpath = os.path.join(targetdir, filename)
+
+                    change_dir(targetdir)
+                    recursive = " --recursive " if recursive else ""
+                    if tag:
+                        cmd = "git clone --branch %s %s %s/%s.git " % (tag, recursive, url, repo_name)
+                    else:
+                        cmd = "git clone %s %s/%s.git" % (recursive, url, repo_name)
+                    (cmdstdouterr, ec) = run_cmd(cmd, log_all=True, log_ok=False, simple=False, regexp=False)
+                    if commit:
+                        change_dir(os.path.join(targetdir, repo_name))
+                        recursive = " && git submodule update " if recursive else ""
+                        cmd = "git checkout %s %s " % (commit, recursive)
+                        (cmdstdouterr, ec) = run_cmd(cmd, log_all=True, log_ok=False, simple=False, regexp=False)
+                        change_dir(targetdir)
+                    cmd = "tar cfvz %s --exclude-vcs %s && rm -rf %s" % (targetpath, repo_name, repo_name)
+                    (cmdstdouterr, ec) = run_cmd(cmd, log_all=True, log_ok=False, simple=False, regexp=False)
+                    return targetpath
```

Comment:
@mboisson Quick review: please move this to a dedicated function, the `obtain_file` method is already too big as is... Throwing that in `filetools` probably makes most sense? And then we also should have a dedicated test for it...

Reply:
Just this or the whole logic to download from git ?

Reply:
I would create a function that takes `git_config` as an argument, and trim this down to something like:

```python
elif git_config:
    return new_function(git_config)
```

(but find a better name than `new_function`, of course...)

Reply:
Ok. Moving it to filetools would require quite a few more parameters, because this bit of code uses a couple of variables internal to easyblock.py : 
- srcpaths[0]
- self.name
it also calls change_dir, I'll try to see how I can rework those.

Reply:
OK, but that's only 2 parameters to add, that's OK.

`change_dir` returns the original dir you were in, you could make sure you restore where you were in the function:

```python
cwd = change_dir(new_dir)
...
change_dir(cwd)

Reply:
done